### PR TITLE
Allow insecure keys for development

### PR DESCRIPTION
**Pull request by sanderpick**
This adds a `secure` flag to API keys, which can be set to false for development purposes. Doing so mean Signature Authentication will be skipped.

### CODE REVIEW

File: api/buckets/client/client_test.go [hunk 1]
```diff
@@ -35,6 +35,37 @@ func TestClient_Init(t *testing.T) {
...
+
+	buck, err := client.Init(ctx, "mybuck")
+	require.Nil(t, err)
+
+	links, err := client.Links(ctx, buck.Root.Key)
+	require.Nil(t, err)
+	assert.NotEmpty(t, links.URL)
+	assert.NotEmpty(t, links.IPNS)
+}
+
```
Comment (sanderpick) [inline api/buckets/client/client_test.go:68]: Just adding these as they were missing from coverage.

File: cmd/hub/cli/keys.go [hunk 2]
```diff
@@ -53,19 +54,33 @@ API secrets should be kept safely on a backend server, not in publicly readable
...
 		}
-		index, keyType, err := prompt.Run()
+		index, keyTypeDesc, err := prompt.Run()
 		if err != nil {
 			cmd.End("")
 		}
 
+		var domains []string
+		keyType := pb.KeyType(index)
+		if keyType == pb.KeyType_USER {
```
Comment (sanderpick) [inline cmd/hub/cli/keys.go:64]: If it's a user key, prompt for optional allowed origin domains

File: collections/ffs.go [hunk 3]
```diff
@@ -8,31 +8,31 @@ import (
 )
 
 type FFSInstance struct {
-	BucketKey string `bson:"_id"`
-	FFSToken  string
-	Archives  Archives
+	BucketKey string   `bson:"_id"`
+	FFSToken  string   `bson:"ffs_token"`
```
Comment (sanderpick) [inline collections/ffs.go:12]: Adding tags here that will make the raw keys match the format of everything else, which is nice when using the `mongo` shell to explore data.

File: core/core.go [hunk 4]
```diff
@@ -421,18 +422,22 @@ func (t *Textile) authFunc(ctx context.Context) (context.Context, error) {
...
+		if !ok {
+			origin, ok := common.OriginFromMD(ctx)
+			if !(ok && key.AllowsOrigin(origin)) {
+				return nil, status.Error(codes.Unauthenticated, "API key signature required")
+			}
+		} else {
+			ctx = common.NewAPISigContext(ctx, msg, sig)
+			if !common.ValidateAPISigContext(ctx, key.Secret) {
+				return nil, status.Error(codes.Unauthenticated, "Bad API key signature")
+			}
```
Comment (sanderpick) [inline core/core.go:440]: Here's the actual auth check.

File: api/buckets/pb/buckets.proto [hunk 5]
```diff
@@ -4,7 +4,7 @@ package buckets.pb;
 option java_multiple_files = true;
 option java_package = "io.textile.buckets_grpc";
 option java_outer_classname = "TextileBuckets";
-option objc_class_prefix = "TXTL_BUCKETS";
+option objc_class_prefix = "TT_BUCKETS";
```
Comment (sanderpick) [inline api/buckets/pb/buckets.proto:7]: Just cleaning these up.

File: cmd/hub/cli/keys.go [hunk 6]
```diff
@@ -53,19 +55,27 @@ API secrets should be kept safely on a backend server, not in publicly readable
...
 		}
-		index, keyType, err := prompt.Run()
+		index, keyTypeDesc, err := prompt.Run()
 		if err != nil {
 			cmd.End("")
 		}
 
+		var secure bool
+		promptSecure := promptui.Prompt{
+			Label:     "Require Signature Authentication (recommended)",
```
Comment (sanderpick) [inline cmd/hub/cli/keys.go:65]: Prompt for sig auth

File: core/core.go [hunk 7]
```diff
@@ -421,18 +422,21 @@ func (t *Textile) authFunc(ctx context.Context) (context.Context, error) {
...
-		}
-		ctx = common.NewAPISigContext(ctx, msg, sig)
 		key, err := t.collections.APIKeys.Get(ctx, k)
 		if err != nil || !key.Valid {
 			return nil, status.Error(codes.NotFound, "API key not found or is invalid")
 		}
-		if !common.ValidateAPISigContext(ctx, key.Secret) {
-			return nil, status.Error(codes.Unauthenticated, "Bad API key signature")
+		ctx = common.NewAPIKeyContext(ctx, k)
+		if key.Secure {
```
Comment (sanderpick) [inline core/core.go:430]: Just skip if key is not secure.

File: api/hub/pb/hub.proto [hunk 8]
```diff
@@ -39,6 +39,7 @@ message GetSessionInfoReply {
 
 message CreateKeyRequest {
     KeyType type = 1;
+    bool secure = 2;
```
Comment (jsign) [inline api/hub/pb/hub.proto:42]: 👍

File: collections/ffs.go (hunk 3 above)
Comment (jsign) [reply collections/ffs.go:12]: Cool.

**Pull request merged**